elastic/beats
# Kubernetes metadata enhancements

## What does this PR do?

1.  adds `kubernetes.host.hostname` on node metadata, 
2. attach node metadata on pod metadata by default (without `add_resource_metadata` required to enabled) for both events coming from autodiscovered pods but also for events enriched by `add_kubernetes_metadata`
3. attach namespace metadata on pod metadata by default (without `add_resource_metadata` required to enabled) for both events coming from autodiscovered pods but also for events enriched by `add_kubernetes_metadata`
4. Enables always `add_resource_metadata` section for autodiscover and add_kubernetes_metadata.



## Why is it important?
This information is useful in various ways. See more on the respective issue: https://github.com/elastic/beats/issues/20434


## How to test this PR manually

1. Use autodiscover config like:

### Test with autodiscover provider
```
filebeat.autodiscover:
  providers:
    - type: kubernetes
      node: ${NODE_NAME}
      hints.enabled: true
      hints.default_config:
        type: container
        paths:
          - /var/log/containers/*${data.kubernetes.container.id}.log
```

2. Make sure that Redis module is enabled and that `kubernetes.node.hostname` field is populated.


### Test with add_kubernetes_metadata processor
1. 
```
filebeat.inputs:
- type: container
  paths:
    - /var/log/containers/*.log
  processors:
    - add_kubernetes_metadata:
        host: ${NODE_NAME}
        matchers:
        - logs_path:
            logs_path: "/var/log/containers/"
```
2. Make sure  that `kubernetes.node.hostname` field is populated as well as `kubernetes.node.*` and `kubernetes.namespace.*`.


## Related issues

- Closes https://github.com/elastic/beats/issues/20434

## Screenshots
Filebeat & add_kubernetes_metadata: 


![Screenshot 2020-11-04 at 12 58 08](https://user-images.githubusercontent.com/11754898/98103258-6cdc2e80-1e9d-11eb-9ac8-c97c58d130f3.png)


Autodiscover:


![Screenshot 2020-11-04 at 13 07 30](https://user-images.githubusercontent.com/11754898/98104164-bda05700-1e9e-11eb-96e8-40379e745885.png)
